cmake/TorchConfig.cmake.in

```diff
@@ -51,6 +51,9 @@ list(APPEND TORCH_LIBRARIES ${C10_LIBRARY})
 # shared library.
 # TODO: this list might be incomplete.
 if (NOT @BUILD_SHARED_LIBS@)
+  find_library(TORCH_CPU_LIBRARY torch_cpu PATHS "${TORCH_INSTALL_PREFIX}/lib")
+  list(APPEND TORCH_LIBRARIES ${TORCH_CPU_LIBRARY})
```

Comment:
Hmm... do we need to do CUDA detection here too then? I'm not sure how this config file decides if CUDA is involvd or not. (Also I'm super confused why this cmake is necessary at all in the first place.)

Reply:
We might need do CUDA as well but I don't have environment to test yet so leave a TODO above.

My (naive) understanding is that it is for 3rd party code to use libtorch.so / libtorch*.a in their project.
If they use prebuilt packages then they will get 1) *.so binaries; 2) public headers; 3) public cmakes. I checked this file is in conda package:
```
./python3.7/site-packages/torch/share/cmake/ATen/ATenConfig.cmake
...
./python3.7/site-packages/torch/share/cmake/Torch/TorchConfigVersion.cmake
./python3.7/site-packages/torch/share/cmake/Torch/TorchConfig.cmake
...
./python3.7/site-packages/torch/share/cmake/Caffe2/Caffe2Config.cmake
./python3.7/site-packages/torch/share/cmake/Caffe2/Caffe2ConfigVersion.cmake
./python3.7/site-packages/torch/share/cmake/Caffe2/Caffe2Targets.cmake
...
./python3.7/site-packages/torch/share/cmake/Caffe2/public/gflags.cmake
./python3.7/site-packages/torch/share/cmake/Caffe2/public/protobuf.cmake
```

We don't expose our regular cmake as people don't have access to all sources in this case.

For dynamic libraries this TorchConfig.cmake seems working well (I haven't checked lately though). For static libraries it seems to be broken - but I'm not sure if we release any prebuilt static library packages, do we?
This might be useful for some mobile developers - if they want to access c++ API directly and want to strip out dead code at link time, then they can build static libraries on their own and use this TorchConfig.cmake to integrate libtorch into their cmake project. I'm creating a small demo app to show this workflow: https://github.com/ljk53/pytorch-android-cpp-demo/blob/master/CMakeLists.txt

@ezyang does it make sense or I misunderstood something?

Reply:
I think we might have some unadvertised prebuilt static libraries, but it's definitely uncommon. However, I agree it should work.

Your description seems reasonable. What I don't understand is why we have so many different Config.cmake files in the installation. Surely there should just be one? Also, another thing I don't understand is why the dependency on `torch` to `torch_cpu_library` doesn't result in a target that does the right thing at export time. But I guess I should just read the generated cmake and find out for myself!